File: c++/Solution211.cpp
//Trie WordDictionary(): O(1) O(1), addWord(): O(n) O(V), search(): O(nZ) O(V)
//n = word.length(), V為node數量, Z = 26, search()時若遇到'.'需要跑for迴圈O(Z), 總共O(nZ)
class WordDictionary {//Solution211
public:
    vector<WordDictionary*> children;
    bool end_flag;

    WordDictionary() {
        children = vector<WordDictionary*>(26, nullptr);
        end_flag = false;
    }

    void addWord(string word) {
        WordDictionary* root = this;
        for (char ch : word) {
            if (!root->children[ch - 'a']) {
                root->children[ch - 'a'] = new WordDictionary();
            }
            root = root->children[ch - 'a'];
        }
        root->end_flag = true;
    }

    bool search(string word) {
        return search(0, word, this); 
    }

    bool search(int i, string word, WordDictionary* root) {
        if (i == word.length()) return root->end_flag; // 到底了

        char ch = word[i];
        if (ch == '.') {
            for (WordDictionary* child : root->children) {
                if (child && search(i + 1, word, child)) {
                    return true; // 其中有就true
                }
            }
            return false;
        } else {
            if (!root->children[ch - 'a']) {
                return false;
            } else {
                return search(i + 1, word, root->children[ch - 'a']);
            }
        }
    }
};

File: java/Solution211.java
package leetCode.java;

//Trie WordDictionary(): O(1) O(1), addWord(): O(n) O(V), search(): O(nZ) O(V)
//n = word.length(), V為node數量, Z = 26, search()時若遇到'.'需要跑for迴圈O(Z), 總共O(nZ)
class WordDictionary {//Solution211
    WordDictionary[] children;
    boolean isEnd;

    public WordDictionary() {
        children = new WordDictionary[26];
        isEnd = false;
    }
    
    public void addWord(String word) {
        WordDictionary root = this;
        for (char ch : word.toCharArray()) {
            int index = ch - 'a';
            if (root.children[index] == null) {
                root.children[index] = new WordDictionary();
            }
            root = root.children[index];
        }
        root.isEnd = true;
    }
    
    public boolean search(String word) {
        return search(word, 0, this);
    }

    public boolean search(String word, int index, WordDictionary root) {
        if (index == word.length()) {
            return root.isEnd;
        }

        char ch = word.charAt(index);
        if (ch  == '.') {
            for (WordDictionary child : root.children) { // 檢測是否有任何一個child
                if (child != null && search(word, index + 1, child)) {
                    return true; // 只要有一條通就true
                }
            }
            return false;
        } else {
            WordDictionary child = root.children[ch - 'a'];
            if (child == null) {
                return false;
            }
            return search(word, index + 1, child);
        }
    }
}